packages-exp/auth-exp/src/platform_browser/messagechannel/receiver.ts

```diff
@@ -0,0 +1,161 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ReceiverHandler,
+  EventType,
+  ReceiverResponse,
+  SenderMessageEvent,
+  Status,
+  SenderRequest
+} from './index';
+import { _allSettled } from './promise';
+
+/**
+ * Interface class for receiving messages.
+ *
+ * @internal
+ */
+export class Receiver {
+  private static readonly receivers: Receiver[] = [];
+  private readonly boundEventHandler: EventListener;
+
+  private readonly handlersMap: {
+    // Typescript doesn't have existential types :(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [eventType: string]: Set<ReceiverHandler<any, any>>;
+  } = {};
+
+  constructor(private readonly eventTarget: EventTarget) {
+    this.boundEventHandler = this.handleEvent.bind(this);
+  }
+
+  /**
+   * Obtain an instance of a Receiver for a given event target, if none exists it will be created.
+   *
+   * @param eventTarget - An event target (such as window or self) through which the underlying
+   * messages will be received.
+   */
+  static _getInstance(eventTarget: EventTarget): Receiver {
+    // The results are stored in an array since objects can't be keys for other
+    // objects. In addition, setting a unique property on an event target as a
+    // hash map key may not be allowed due to CORS restrictions.
+    const existingInstance = this.receivers.find(receiver =>
+      receiver.isListeningto(eventTarget)
+    );
+    if (existingInstance) {
+      return existingInstance;
+    }
+    const newInstance = new Receiver(eventTarget);
+    this.receivers.push(newInstance);
+    return newInstance;
+  }
+
+  private isListeningto(eventTarget: EventTarget): boolean {
+    return this.eventTarget === eventTarget;
+  }
+
+  /**
+   * Fans out a MessageEvent to the appropriate listeners.
+   *
+   * @remarks
+   * Sends an {@link Status.ACK} upon receipt and a {@link Status.DONE} once all handlers have
+   * finished processing.
+   *
+   * @param event - The MessageEvent.
+   *
+   * @internal
+   */
+  private async handleEvent<
+    T extends ReceiverResponse,
+    S extends SenderRequest
+  >(event: Event): Promise<void> {
+    const messageEvent = event as MessageEvent<SenderMessageEvent<S>>;
+    const { eventId, eventType, data } = messageEvent.data;
+
+    const handlers: Set<ReceiverHandler<T, S>> | undefined = this.handlersMap[
+      eventType
+    ];
+    if (!handlers || handlers.size === 0) {
+      return;
+    }
+
+    messageEvent.ports[0].postMessage({
+      status: Status.ACK,
+      eventId,
+      eventType
+    });
+
+    const promises = Array.from(handlers).map(async handler => {
```

Comment:
this can be simplified:
```ts
const promises = Array.from(handlers).map(handler => handler(messageEvent.origin, data));
```

Reply:
we need the async function convert from `Promise<foo> | foo` -> `Promise<foo>`

Reply:
ah. well still it can be `Array.from(handlers).map(async handler => handler(messageEvent.origin, data));`

Reply:
done